### DISCUSSION TITLE
Scripting

### DISCUSSION THREAD

**Discussion by msspad** [sn4k3/UVtools#473] [General]
Some minor issues I found:
- ScriptInit() is run twice, surprising but no big issue
- After clicking on the button "Scripting", the modal message "Are you sure..." pops up. It's somewhat anoying to confirm it every time. Could you remove it? Same is true for the ctb-encryption warning that pops up every time.
- Window size for full screen is not working with a 2 monitor setup on Windows 10?
- It's usefull to understand the different rectangles for writing scripts. Maybe you could add to the method documentation something along the lines:
-- Operation.OriginalBoundingRectangle is the bounding rectangle of all pixels on all layers
-- layer.BoundingRectangle is the bounding rectangle of the pixels on a single layer (=not all layers)
-- Operation.ROI is the rectangle created by the UVTools user with Shift+mouse dragging. Is only defined when user created it.

**Comment by sn4k3** [comment]
> ScriptInit() is run twice, surprising but no big issue

My test (a breakpoint on ScriptInit() + a debug message) shows it trigger only once. A search on solution reveal only one usage call.  
Where are you getting this behaviour?

> After clicking on the button "Scripting", the modal message "Are you sure..." pops up. It's somewhat anoying to confirm it every time. Could you remove it? Same is true for the ctb-encryption warning that pops up every time.

All tools have confirmation dialog, they wont get removed, but i can introduce a setting so user can opt for no confirmation.
E-CTB message will remain for now

> Window size for full screen is not working with a 2 monitor setup on Windows 10?

Need more explanation on this, screenshots.
But programs you set full screen will only do it on current (1) monitor, not both.


> It's usefull to understand the different rectangles for writing scripts

The names are self-explanatory, but I can add the missing documentation over them. 

`SlicerFile.BoundingRectangle` is the actual bounds and can change during execution
`Operation.OriginalBoundingRectangle` is the bounds before operation executed and does not change during execution when a mat has been changed

**Reply by msspad** [reply]
> > ScriptInit() is run twice, surprising but no big issue
> 
> My test (a breakpoint on ScriptInit() + a debug message) shows it trigger only once. A search on solution reveal only one usage call. Where are you getting this behaviour?
> 

Correct, it runs now only once. A few weeks ago, I verified it with logging but it does not occur anymore. Sorry for the fuss.



> > After clicking on the button "Scripting", the modal message "Are you sure..." pops up. It's somewhat anoying to confirm it every time. Could you remove it? Same is true for the ctb-encryption warning that pops up every time.
> 
> All tools have confirmation dialog, they wont get removed, but i can introduce a setting so user can opt for no confirmation. E-CTB message will remain for now

An option for no confirmations would be great!



> 
> > Window size for full screen is not working with a 2 monitor setup on Windows 10?
> 
> Need more explanation on this, screenshots. But programs you set full screen will only do it on current (1) monitor, not both.

Well, I close UVTools like this (2 monitors):
![image](https://user-images.githubusercontent.com/6191764/168140923-7a0637d6-8cfa-4cd3-b3ac-54f1b3314d65.png)

and it opens like this when I double-click on a .ctb file:
![image](https://user-images.githubusercontent.com/6191764/168140322-96e1c31c-c020-4bd2-af5e-dc0c0146d782.png)



> 
> > It's usefull to understand the different rectangles for writing scripts
> 
> The names are self-explanatory, but I can add the missing documentation over them.
> 
> `SlicerFile.BoundingRectangle` is the actual bounds and can change during execution `Operation.OriginalBoundingRectangle` is the bounds before operation executed and does not change during execution when a mat has been changed

ok, so we have those bounds:
- Rectangle SlicerFile.BoundingRectangle
- Rectangle Operation.ROI
- Rectangle Operation.OriginalBoundingRectangle
- Mat Operation.GetRoiOrVolumeBounds(layerMat)
- Rectangle layer.BoundingRectangle

I guess you use "Volume" in the method name `Operation.GetRoiOrVolumeBounds(layerMat)` because it returns the OpenCV mat of the layer?

**Reply by sn4k3** [reply]
Check settings, does the start up is set to maximized?

VolumeBounds is an alias for BoundingRectangle, that includes union of all layers bounds, in other word is the `OriginalBoundingRectangle`, but is not called BoundingRectangle there because ROI is also an "Bounding rectangle", thats why the word VolumeBounds.

That method returns a mat cropped by ROI if availiable, otherwise cropped by model bounds.
https://github.com/sn4k3/UVtools/blob/51e4f329879298e589ce7f5cc1e0cd2db64c5394/UVtools.Core/Operations/Operation.cs#L484

Useful to relief some processing, imagine that volume bounds in a 8k file is just 100x100, "cropping" Mat to that size will speed up the arithmetic over the Mat, eg: sum 5000x5000 pixels is slower than sum 100x100.
As a ROI is just an advance of a pointer (Mat.IntPtr + first pixel of roi) + a offset per row, it have "no cost" "cropping" the mat when needed and memory is shared and hold by original Mat, when you create a mat from a ROI you are not allocating nor creating new image, memory is hold by the parent/original Mat. UVtools utility methods are very convenient.

**Reply by msspad** [reply]
I have set it like this 
![image](https://user-images.githubusercontent.com/6191764/168222896-fa600b0f-598a-4d9b-8bcb-2659c4a04b69.png)
because I change forth and back with maximized/non-maximized windows. When the option maximized is active, the window get maximized reliably.

Thanks for BoundingRectangle explanation, I'm also a happy user of ROI. I was wondering whether the ROI is reduced to the `Operation.OriginalBoundingRectangle`. When manually marking the ROI, I typically draw the ROI over the bounds of `Operation.OriginalBoundingRectangle` and it would save same more calculation time.

**Reply by msspad** [reply]
I forgot to add that "Start maximized" does not restore the window in the screen where UVTools was last closed.

**Reply by sn4k3** [reply]
I will test that under two monitors, but i suspect when you maximize the size event dont fire. Need to check it out.

> I forgot to add that "Start maximized" does not restore the window in the screen where UVTools was last closed.

Thats true, when that is checked none of the other are take into consideration, maybe it should? https://github.com/sn4k3/UVtools/blob/51e4f329879298e589ce7f5cc1e0cd2db64c5394/UVtools.WPF/App.axaml.cs#L229

> I'm also a happy user of ROI.

There's also Masks (Shift + Alt + click), it select masses instead. If you right click on ROI button on UVtools GUI you will see some possible auto selections (presets). 

![UVtools_2022-05-13_14-36-07](https://user-images.githubusercontent.com/113281/168295402-3323406e-08c2-4b5c-91e3-debf916fae2b.png)

Using masks is a bit more laborous, you need to clone the original mat and keep it intact, then modify the target, in the end call Operation.ApplyMask(original, target) then UVtools will transpose the target modifications to original and set original to target keeping only modifications inside the mask selection. This way you can avoid holes but it require extra effort ofc.

![image](https://user-images.githubusercontent.com/113281/168298254-c04ccfd9-99ab-40ec-a1a4-a90285c75a88.png)


> When manually marking the ROI, I typically draw the ROI over the bounds of Operation.OriginalBoundingRectangle

`Ctrl + Shift + B` 😉

**Reply by msspad** [reply]
> I will test that under two monitors, but i suspect when you maximize the size event dont fire. Need to check it out.
> 
> > I forgot to add that "Start maximized" does not restore the window in the screen where UVTools was last closed.
> 
> Thats true, when that is checked none of the other are take into consideration, maybe it should?

I obviously can only speak for myself: I would like to have the window open at the same location and same size as when it was last closed (no matter whether maximized or not, no matter on which screen).



> > I'm also a happy user of ROI.
> 
> There's also Masks (Shift + Alt + click), it select masses instead. If you right click on ROI button on UVtools GUI you will see some possible auto selections (presets).
> 

> 
> Using masks is a bit more laborous, you need to clone the original mat and keep it intact, then modify the target, in the end call Operation.ApplyMask(original, target) then UVtools will transpose the target modifications to original and set original to target keeping only modifications inside the mask selection. This way you can avoid holes but it require extra effort ofc.
> 


Yes, masks are great, I have used them when applying standard operations in UVTools. In scripts, I prefer to do it directly with OpenCV, also because code can apply some stuff automatically without user intervention.


> 
> > When manually marking the ROI, I typically draw the ROI over the bounds of Operation.OriginalBoundingRectangle
> 
> `Ctrl + Shift + B` 😉

Didn't know that, thanks. I meant more in scripting, when getting the `Operation.ROI` Rectangle, it is not capped at `Operation.OriginalBoundingRectangle`, right?

**Reply by sn4k3** [reply]
> I obviously can only speak for myself: I would like to have the window open at the same location and same size as when it was last closed (no matter whether maximized or not, no matter on which screen).

I can patch to that.

> when getting the Operation.ROI Rectangle, it is not capped at Operation.OriginalBoundingRectangle, right?

Correct, ROI is as user selected, is a rectangle which can be empty (0,0,0,0 == !HaveROI) in case of no selection. It won't fall back to VolumeBounds (Which is what GetRoiOrVolumeBounds() does)

**Reply by sn4k3** [reply]
Using same settings as you with two monitors works to me after maximizing, close, reopen, produces: 
![2022-05-13_21-58-05](https://user-images.githubusercontent.com/113281/168388755-f6100394-d15b-4fc6-9de8-f9420d536991.png)

**Reply by msspad** [reply]
My display 1 is scaled at 125%. When I turn it back to 100%, the window behaves as expected. But with 100%, the fonts are too small. Could C#/UVTools scaled displays correctly?

![image](https://user-images.githubusercontent.com/6191764/168414887-4d4b43e5-51d2-463a-9a29-c76a3aa7ddbe.png)

**Reply by sn4k3** [reply]
Ah scaling... Thats always a problem.
The UI framework still have some bugs regarding scaling. Is out of my scope.
